crates/autopilot/src/domain/competition/participation_guard/statistics.rs

```diff
@@ -0,0 +1,511 @@
+    banned_solvers: dashmap::DashMap<eth::Address, u32>,
+    /// Stop banning solvers when the number of remaining active
+    /// solvers hits this threshold. This is crucial to prevent the protocol
+    /// from blocking itself by banning all the solvers since the recover
+    /// mechanism requires new competitions.
+    min_active_solvers_threshold: u32,
+    config: StatisticsBasedSolverParticipationGuardConfig,
+    drivers_by_address: HashMap<eth::Address, Arc<infra::Driver>>,
+    competitions_tracker: Mutex<CompetitionsTracker>,
+}
+
+/// Keeps track of the solver's competition submission statistics.
+#[derive(Clone, Debug, Default)]
+struct SolverCompetitionStats {
+    total: u32,
+    failed: u32,
+}
+
+impl SolverCompetitionStats {
+    fn increment(&mut self, success: bool, high_failure_threshold: f64) {
+        if success {
+            self.total += 1;
+            // Recover fast in case the solver is banned.
+            if self.failed > 0 {
+                // Keep the failure rate at the threshold level, so the solver is
+                // able to participate again. And at the same time, on the next failure it gets
+                // banned again.
+                while self.failure_rate() >= high_failure_threshold {
+                    self.total += 1;
+                }
+            }
+        } else if !success && self.failure_rate() <= high_failure_threshold {
+            // Keep the failure rate at the same level, so the solver is
+            // able to recover fast with a single successful settlement.
+            self.failed += 1;
+            self.total += 1;
+        }
+    }
+
+    fn decrement(&mut self, success: bool) {
+        self.total = self.total.saturating_sub(1);
+        if !success {
+            self.failed = self.failed.saturating_sub(1);
+        }
+    }
+
+    fn failure_rate(&self) -> f64 {
+        if self.total == 0 {
+            0.0
+        } else {
+            self.failed as f64 / self.total as f64
+        }
+    }
+}
+
+/// Keeps track of the last N competitions metadata to find solvers with high
+/// failure rates and consecutive failed settlements.
+/// Keeping only the last N competitions metadata is considered more efficient
+/// for solvers to recover from a temporary failure state.
+struct CompetitionsTracker {
+    /// A FIFO queue to store the last N competitions metadata.
+    queue: VecDeque<competition::Metadata>,
+    /// The maximum size of the queue where the oldest metadata will be removed.
+    max_cache_size: usize,
+    /// The threshold to consider the solver as a low-settling solver. Once the
+    /// threshold is reached, it stops growing on consecutive failed settlements
+    /// to make is possible for solver to recover with a single successful
+    /// settlement.
+    high_failure_threshold: f64,
+    /// The minimum number of competitions the solver should win to be
+    /// considered as low-settling.
+    min_won_competitions: u32,
```

Comment:
The doc comment reads a bit weird. I think it's supposed to express that we need at least `n` data points before the mechanism is allowed to kick in but it reads like winning competitions makes them bad solvers that need to get punished.